ath79: add support for WNDR3700 and WNDR3700v2

hnyman opened the pull request:
@mkresin 
Thanks for advice regarding the .dtsi / .dts formulation. I left only the partitions in the .dts

Add support for WNDR3700 and WNDR3700v2

The three routers are identical except
  * device IDs
  * WNDR3700 (v1) has only 8 MB flash, while others have 16 MB. Partition structure needs to be defined for each device.
  * (Additionally, WNDR3800 has 128 MB RAM, but RAM size is not in DTS)

Commits:
* separate the common parts from old wndr3800.dts into wndr3700.dtsi and leave just the device-specific things into wndr3800.dts
* add definitions for WNDR3700 and WNDR3700v2

Note: wifi related things have been left out of this PR to keep this simple.

Tested with WDNR3800 and WNDR3700v2

hnyman committed:
ath79: create WNDR3700 series .dtsi and adjust WNDR3800

Prepare for addition of WNDR3700 and WNDR3700v2 by
separating the common parts into wndr3700.dtsi and
leaving just the device-specific things into wndr3800.dts

The three routers are identical except
  * device IDs
  * WNDR3700 (v1) has only 8 MB flash, while others have 16 MB.
    Partition structure needs to be defined for each device.
  * (WNDR3800 has 128 MB RAM, but RAM size is not in DTS)

Also separate the common parts of the image recipe.
(Drop also the initramfs recipe.)

Signed-off-by: Hannu Nyman <hannu.nyman@iki.fi>

xdarklight commented on target/linux/ath79/dts/ar7161_netgear_wndr3700.dts:
```diff
@@ -0,0 +1,38 @@
+		reg = <0x070000 0x780000>;
+	};
+
+	art: partition@3 {
```
it seems that it was wrong before, but the unit-address (the number after the "@") should always match the address cell(s)
thus the correct value would be 7f0000 in this case (or in other words: use partition@7f0000 instead of partition@3)

the same comment also applies to all other partitions except the u-boot one

hnyman replied:
Based on DTS specs, you are right. I fixed that.

(Likely it is mostly cosmetics, though, as it has been this 0 1 2 3 all the time in the WNDR3800 DTS file, and it has worked)

mkresin commented:
@hnyman would you please move common image build code use by all three boards into a shared template.

Is it correct that neither the wdr3700 nor the wdr3700v2 requires any userspace config? something like network, led and so on.

If I get it right, this PR should be merged prior to PR#1258. Is it correct?

hnyman commented:
Yes, this should be merged first.  I will rewrite #1258 wifi PR to serve all three routers (from the .dtsi)

Thanks for pointing out the user_space things. I had the 02_network interface/switch addition for WNDR3700+v2 there in my own testing, but I accidentally removed that at the same time as I cleaned out the wifi parts for this PR. I have re-added that. 

LEDs come from DTS like for WNDR3800, so user-space there.

> would you please move common image build code use by all three boards into a shared template.

You mean similarly, as in AR71xx?  

I tried that earlier and failed, but that was before you fixed DEVICE_VARS a few days ago with 38877d88
I will try again.

mkresin commented:
> You mean similarly, as in AR71xx?

Yes similar but not identical to.

The way it is done in ar71xx is suboptimal as images including the Device/wndr3700 are overwriting parts of it. Which can cause hard to find issues if something is defined in the template but not required or required in a different way for an image including the template. b94b4b0862daeae635285519f5eed2495165b1b5 would be such an example.

I've something like the following in mind:

```
define Device/netgear_wndr3x00
  ATH_SOC := ar7161
  KERNEL := kernel-bin | append-dtb | lzma -d20 | netgear-uImage lzma
  IMAGES := sysupgrade.bin factory.img
  IMAGE/default := append-kernel | pad-to $$$$(BLOCKSIZE) | netgear-squashfs | append-rootfs | pad-rootfs
  IMAGE/sysupgrade.bin := $$(IMAGE/default) | append-metadata | check-size $$$$(IMAGE_SIZE)
  IMAGE/factory.img := $$(IMAGE/default) | netgear-dni | check-size $$$$(IMAGE_SIZE)
  DEVICE_PACKAGES := kmod-usb-core kmod-usb-ohci kmod-usb2 kmod-usb-ledtrig-usbport kmod-leds-reset
endef

define Device/netgear_wndr3700
  $(Device/netgear_wndr3x00)
  DEVICE_TITLE := NETGEAR WNDR3700
  NETGEAR_KERNEL_MAGIC := 0x33373030
  NETGEAR_BOARD_ID := WNDR3700
  IMAGE_SIZE := 7680k
  IMAGES += factory-NA.img
  IMAGE/factory-NA.img := $$(IMAGE/default) | netgear-dni NA | check-size $$$$(IMAGE_SIZE
  SUPPORTED_DEVICES += wndr3700
endef
TARGET_DEVICES += netgear_wndr3700

define Device/netgear_wndr3700v2
  $(Device/netgear_wndr3x00)
  DEVICE_TITLE := NETGEAR WNDR3700v2
  NETGEAR_KERNEL_MAGIC := 0x33373031
  NETGEAR_BOARD_ID := WNDR3700v2
  NETGEAR_HW_ID := 29763654+16+64
  IMAGE_SIZE := 15872k
  SUPPORTED_DEVICES += wndr3700v2
endef
TARGET_DEVICES += netgear_wndr3700v2

define Device/netgear_wndr3800
  $(Device/netgear_wndr3x00)
  DEVICE_TITLE := NETGEAR WNDR3800
  NETGEAR_KERNEL_MAGIC := 0x33373031
  NETGEAR_BOARD_ID := WNDR3800
  NETGEAR_HW_ID := 29763654+16+128
  IMAGE_SIZE := 15872k
  SUPPORTED_DEVICES += wndr3800
endef
TARGET_DEVICES += netgear_wndr3800
```

I dropped the KERNEL_INITRAMFS as well. It defaults to the KERNEL recipe anyway.

hnyman commented:
Well, I already wrote it like in ar71xx and am currently compiling it...
But I will rewrite it. You are right that the ar71xx version can be confusing as the same variable can be modified three times.

This was the ar71xx-like version currently being compiled:
```
define Device/netgear_wndr3700
  ATH_SOC := ar7161
  DEVICE_TITLE := NETGEAR WNDR3700
  NETGEAR_KERNEL_MAGIC := 0x33373030
  KERNEL := kernel-bin | append-dtb | lzma -d20 | netgear-uImage lzma
  KERNEL_INITRAMFS := kernel-bin | append-dtb | lzma -d20 | netgear-uImage lzma
  NETGEAR_BOARD_ID := WNDR3700
  IMAGE_SIZE := 7680k
  IMAGES := sysupgrade.bin factory.img factory-NA.img
  IMAGE/default := append-kernel | pad-to $$$$(BLOCKSIZE) | netgear-squashfs | append-rootfs | pad-rootfs
  IMAGE/sysupgrade.bin := $$(IMAGE/default) | append-metadata | check-size $$$$(IMAGE_SIZE)
  IMAGE/factory.img := $$(IMAGE/default) | netgear-dni | check-size $$$$(IMAGE_SIZE)
  IMAGE/factory-NA.img := $$(IMAGE/default) | netgear-dni NA | check-size $$$$(IMAGE_SIZE)
  DEVICE_PACKAGES := kmod-usb-core kmod-usb-ohci kmod-usb2 kmod-usb-ledtrig-usbport kmod-leds-reset
  SUPPORTED_DEVICES += wndr3700
endef
TARGET_DEVICES += netgear_wndr3700

define Device/netgear_wndr3700v2
  $(Device/netgear_wndr3700)
  DEVICE_TITLE := NETGEAR WNDR3700v2
  NETGEAR_KERNEL_MAGIC := 0x33373031
  NETGEAR_BOARD_ID := WNDR3700v2
  NETGEAR_HW_ID := 29763654+16+64
  IMAGE_SIZE := 15872k
  IMAGES := sysupgrade.bin factory.img
  SUPPORTED_DEVICES += wndr3700v2
endef
TARGET_DEVICES += netgear_wndr3700v2

define Device/netgear_wndr3800
  $(Device/netgear_wndr3700v2)
  DEVICE_TITLE := NETGEAR WNDR3800
  NETGEAR_BOARD_ID := WNDR3800
  NETGEAR_HW_ID := 29763654+16+128
  SUPPORTED_DEVICES += wndr3800
endef
TARGET_DEVICES += netgear_wndr3800
```

hnyman committed:
ath79: add WNDR3700 and WNDR3700v2

Add support for WNDR3700 and WNDR3700v2.

They share most things with WNDR3800.
Only device IDs and partition structure needs to be set.

Note: WNDR3700 (v1) has no NETGEAR_HW_ID, but has
      also the NA version of the factory image.

Signed-off-by: Hannu Nyman <hannu.nyman@iki.fi>

hnyman commented:
@mkresin 

I combined the common image build code, as discussed.

I did some git magic and there are now just two commits: 
* first splits the old 3800 into common & device-specific parts.
* second adds 3700 and 3700v2

I have rebased on the current HEAD.

mkresin commented:
merged, thanks!
